File: day3a.py

with open("day3.txt") as f:
    counter = 0
    for line in f:
        line = line.strip()
        length = len(line)
        bag1 = set(line[0:length//2])
        bag2 = set(line[length//2:length])

        # print(bag1, bag2)
        alphebet = 0
        for alphebet in bag1:
            if alphebet in bag2:
                if ord(alphebet) >= 96:
                    score = ord(alphebet) - 96
                else:
                    score = ord(alphebet) - 38
                # print(alphebet, score)
                counter += score

    print(counter)

File: src/bin/day3a.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::collections::HashSet;

fn main() -> io::Result<()> {
    let file = File::open("day3.txt")?;
    let mut total: i32 = 0;
    let scores = " abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".as_bytes();
    for line in io::BufReader::new(file).lines() {
        let line = line?;
        let a: HashSet<_> = line[..line.len()/2].as_bytes().iter().cloned().collect();
        let b: HashSet<_> = line[line.len()/2..].as_bytes().iter().cloned().collect();
        for c in a.intersection(&b) {
            let score = scores.iter().position(|&i| i == *c).unwrap() as i32;
            total += score;
            // println!("{} {} {}", c, score, total);
        }
    }
    println!("{}", total);
    Ok(())
}
